### DISCUSSION TITLE
`$permissions`

### DISCUSSION THREAD

**Discussion by uBlock-user** [uBlockOrigin/uBlock-issues#2714] [General] [answered]
@gorhill Seems AdGuard has implemented permissions policy modifier via [$permissions](https://adguard.com/kb/general/ad-filtering/create-own-filters/#permissions-modifier), did they do anything different than  https://github.com/gorhill/uBlock/commit/af6c992d69c0d0ff5637b4b34b4210cfa31e7127 ?

**Comment by gorhill** [comment]
Forgot completely about that commit. I could implement this in the current dev cycle but not sure there are AdGuard lists using it, not worth it if not.

**Reply by uBlock-user** [reply]
So does it work after all ?

**Reply by gorhill** [reply]
No since that commit was in a branch which was never merged.

**Reply by uBlock-user** [reply]
I meant when you tested the commit in your local build...

**Reply by gorhill** [reply]
Sorry, I can't remember, I had even forgotten I worked on this.

**Reply by uBlock-user** [reply]
https://github.com/uBlockOrigin/uBlock-issues/issues/2404#issuecomment-1373958551

**Reply by mtxadmin** [reply]
Came here from [wiki](https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#permissions). So, some privacy-gathering sites can be restricted? At last!

Is it possible to disable queries to Location and Notifications functionality on a site-basis level? I suppose, it will be
```
example.com$permissions=geolocation()
```
But this rule is marked red in the editor, and doesn't work. Rule from wiki is red too
```
||example.com^$permissions=browsing-topics=()
```

**Reply by stephenhawk8054** [reply]
@mtxadmin Does this work?

```adb
||example.com^$permissions=geolocation=()
```

**Reply by gorhill** [reply]
> Rule from wiki is red too

What version of uBO are you using? `permissions=` is currently only in ~~devbuild.~~ 1.51.0+.

**Reply by stephenhawk8054** [reply]
> What version of uBO are you using? `permissions=` is currently only in devbuild.

I thought it's in 1.51.0 (stable release) now? Mine doesn't show red in 1.51.0: https://ttm.sh/IUU.png

**Reply by gorhill** [reply]
Right, I corrected myself.

**Reply by gorhill** [reply]
> `||example.com^$permissions=geolocation()`

This won't work, the documentation for `Permissions-Policy` says that we have to use `=`:

    ||example.com^$permissions=geolocation=()

uBO does not have a permission-policy parser, so it won't warn if there are errors in the policy directives themselves.

If the line is rendered as an error, this means there is a filter syntax error, which is what would happen when using an older version of uBO.

**Reply by stephenhawk8054** [reply]
Ah right, I forgot that `=`

**Reply by mtxadmin** [reply]
uBlock Origin updated to 1.51.0 now:
- red "errors" in the editor are gone
- location requests though are not blocking. 

I found simple but useful site for testing and used the forms:
mylocation.org$permissions=geolocation=()
||mylocation.org$permissions=geolocation=()
||mylocation.org^$permissions=geolocation=()

**Reply by gorhill** [reply]
> location requests though are not blocking.

I don't see the site using geolocation API. Here is a site which does query geolocation API: `https://whatmylocation.com/`, `whatmylocation.com$permissions=geolocation=()` works as expected.

**Reply by stephenhawk8054** [reply]
I think it wraps `navigator.geolocation` to another variable `geoPosition`?

The script is `https://mylocation.org/js/geoPosition.min.js`

**Reply by gorhill** [reply]
With no filter, my browser is set to ask whether to allow the site to see my location. The browser does not ask, meaning the API is not being used. The site simply uses IP address information to infer location.

---

I see that `https://mylocation.org/` has a "Browser Geolocation" test button. It fails with `mylocation.org$permissions=geolocation=()`:

![Screenshot from 2023-08-21 11-56-19](https://github.com/uBlockOrigin/uBlock-issues/assets/585534/bd5f230c-0ffe-4664-8ac6-0f8f4435aeef)

**Reply by stephenhawk8054** [reply]
I think you mean the `Public IP Address` section? There's a button below named `Browser Geolocation` -> Click on that -> Click on `Start Test` -> Browser will ask for location permission

**Reply by stephenhawk8054** [reply]
Hmmm... weird, my Firefox still asks for permission even when applying the filter: https://github.com/uBlockOrigin/uBlock-issues/assets/66517106/82dfad49-04d0-4192-a7b4-29e35d7d8f85

**Reply by gorhill** [reply]
For `geolocation` on Firefox, the documentation says partial support: <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Permissions-Policy#browser_compatibility>:

> Only supported through the `allow` attribute on `<iframe>` elements.

**Reply by uBlock-user** [reply]
Yeah it doesn't work in Firefox and Nightly.

**Reply by stephenhawk8054** [reply]
Ah I see

**Reply by gorhill** [reply]
Anyways, the header is injected in Firefox, if the browser does not interpret it, or interpret it in a different way that someone want it to be interpreted, please open an issue with Firefox devs.

---

Meta bugzilla entry for `Permissions-Policy`: <https://bugzilla.mozilla.org/show_bug.cgi?id=permissions-policy>

**Reply by uBlock-user** [reply]
@gorhill header not yet implemented as per - https://bugzilla.mozilla.org/show_bug.cgi?id=1694922

**Comment by gorhill** [comment, marked as answer]
https://github.com/gorhill/uBlock/commit/5ebdbf3e24393560156fdcd931e31f901471f7f3

**Reply by stephenhawk8054** [reply]
Oh, didn't expect Topics is part of permissions. Would it be good if we use `*$permissions=browsing-topics=()` for chrome browsers like with FLoC?

**Reply by gorhill** [reply]
Sure, we should discuss whether we should add this by default, or whether we should have a checkbox in the dashboard.

**Reply by uBlock-user** [reply]
Does it work in Firefox without enabling the feature via `about:config` ?

**Reply by gorhill** [reply]
uBO does not treat the option any differently in any browser. I don't know whether Firefox might complain that the response header is not recognized.

---

Fixed typo: added _not_

**Reply by MasterKia** [reply]
https://github.com/uBlockOrigin/uBlock-issues/wiki/Static-filter-syntax#permissions

However I couldn't find the two examples in the commit ( `oversized-images` and `unsized-media`) on MDN:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Permissions_Policy#browser_compatibility

**Reply by uBlock-user** [reply]
Autoplay still cannot be blocked with the filter, hopefully this changes in future.

**Reply by gorhill** [reply]
> couldn't find the two examples in the commit

I copied the example from AdGuard's documentation, just to show the difference on how to separate multiple directives.

There is this site which mentions those directives: <https://docs.w3cub.com/http/headers/feature-policy#syntax>. I don't know how reliable.

**Comment by gorhill** [comment]
@stephenhawk8054 Gave more thoughts to this and I think we should add it to _uBlock filters -- Privacy_ list. I don't think a user interface component for this is the way to go given that so far it's an API supported only by one browser, and that the API is to serve advertisers in the first place.

**Reply by stephenhawk8054** [reply]
Yeah, I don't think we need to complicate the code more just for that option.

**Reply by stephenhawk8054** [reply]
However, do we have any ways to confirm if it works for a site?

**Reply by gorhill** [reply]
I tried it after enabling _Privacy Sandbox Ads APIs_ in `chrome://flags/` and confirmed this works:

1. Enable _Privacy Sandbox Ads APIs_ in `chrome://flags/`, restart browser
2. Visit `https://arstechnica.com/`
3. Open dev console for page, try `document.browsingTopics()`
4. Result: promise successfully fulfilled
5. Add `*$permissions=browsing-topics=()`
6. Force a reload of `https://arstechnica.com/`
7. Open dev console for page, try `document.browsingTopics()`
8. Result: `the "browsing-topics" Permissions Policy denied the use of document.browsingTopics()`

**Reply by stephenhawk8054** [reply]
First report of breakage for `*$permissions=browsing-topics=()` from LanikJS on Slack:

> `https://www.cjisonline.com/app/login/expired` is completely blocked unless I allow the page in uBO
> This is a Government site that is used for background verification

This filter works:

```adb
@@||cjisonline.com^$permissions=browsing-topics=()
```

It's not easy to reproduce with this site (needs log in and authentication as far as I see) so would be hard to investigate.

---

Added the fix at the moment: https://github.com/uBlockOrigin/uAssets/commit/c91651c80c574fa42c837be959bac846282131ce

**Reply by MasterKia** [reply]
@gorhill What about blocking the [idle detection API](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Permissions-Policy/idle-detection) in Chromium?

https://mozilla.github.io/standards-positions/#idle-detection-api

```adb
*$permissions=idle-detection=()
```

https://techdows.com/2021/09/disable-chrome-idle-detection.html:
> Mozilla and Apple opposed the API due to privacy concerns, but Google went ahead and released Chrome 94 with it. 

<details>
<summary>Mozilla considered Chrome Idle Detection as harmful, while Apple as negative for Safari.</summary>

> “I consider the Idle Detection API too tempting of an opportunity for surveillance capitalism motivated websites to invade an aspect of the user’s physical privacy, keep longterm records of physical user behaviors, discerning daily rhythms (e.g lunchtime), and using that for proactive psychological manipulations..” **Mozilla Web standards Lead**, Tantek[ said](https://github.com/mozilla/standards-positions/issues/453#issuecomment-888225596).

> Apple has expressed fingerprinting concerns and worried, the Idle Detection API could be used by websites to mine bitcoins.

> “Our concerns are not limited to fingerprinting. There is an obvious privacy concern that this API lets a website observe whether a person is near the device or not. This could be used, for example, to start mining bitcoins when the user is not around or start deploying security exploits, etc.” **Apple’s Ryosuke Niwa** [said](https://lists.webkit.org/pipermail/webkit-dev/2020-October/031562.html) .

</details>

**Reply by gorhill** [reply]
Agreed, let's do this, we have good references to backup the decision.

**Reply by MasterKia** [reply]
https://github.com/uBlockOrigin/uAssets/commit/1987b7f4dcf7b4bdb5dc40016527ec68c03ec343

I checked and it works.

**Reply by stephenhawk8054** [reply]
Is it possible to combine 2 `permissions` filters to 1?

**Reply by MasterKia** [reply]
That would be:
```adb
*$permissions=idle-detection=()|browsing-topics=()
```

But it's not possible to have an exception filter?

```adb
*$permissions=idle-detection=()|browsing-topics=()
@@*$permissions=idle-detection=(),domain=example.com
```

I still see the first filter in the logger and the exception does not work.

Also, I think it should be limited to `$doc` requests only?

**Reply by gorhill** [reply]
Best to not combine them to not break exception filters for either.

When no type is given, uBO will use `document,subdocument` internally when the `permissions` option is used (same as with `csp`.

**Reply by uBlock-user** [reply]
> *$permissions=idle-detection=()|browsing-topics=()

That is a whole filter itself and should not be interpret as  `*$permissions=idle-detection=()` and  `*$permissions=browsing-topics=()`. So, 

a exception filter would work look like this - `@@*$permissions=idle-detection=()|browsing-topics=(),domain=example.org`

**Reply by MasterKia** [reply]
@uBlock-user These do work:
```adb
||google.com^$all
@@||google.com^$doc

||google.com^$script,xhr,image
@@||google.com^$script

*$script,xhr,to=google.com|example.com
@@*$script,to=google.com
```

**Reply by uBlock-user** [reply]
Those are individual network type filters. `idle-detection=()` is a policy not a network type filter. Here's there's only one network type, that is  `$permissions`.

**Reply by gorhill** [reply]
> These do work

uBO does not have a permission policy value parser, and neither a CSP value parser, hence why the values can only be wholly excepted. 

This has always been the way for CSP value, there was never a way to except only the `script-src` part of a filter such as `*$csp=worker-src 'none';script-src 'none'`, and unsurprisingly it's also the case for `permissions=` option.

**Reply by stephenhawk8054** [reply]
I guess we can add `Attribution Reporting`: https://developer.chrome.com/docs/privacy-sandbox/attribution-reporting/ ?

```
Measure when an ad click or view leads to a conversion, such as a purchase on an advertiser site.
```

```adb
*$permissions=attribution-reporting=(),from=~localhost
```

**Comment by Yuki2718** [comment]
BTW it will be convenient if we can exclude only these permission rules from the logger. As they're always applied, pretty much are noise.

**Reply by MasterKia** [reply]
Yes, should be disabled by default in the logger.

**Reply by gorhill** [reply]
https://github.com/gorhill/uBlock/commit/a9ff41ae5515d571cf988bc508d9c85a91b52b3b

**Reply by gorhill** [reply]
I can make "not permissions" be default filter expression if you majority agrees.

**Reply by ghost** [reply]
Save the status? How it's done in popup?

**Reply by gorhill** [reply]
Too much code to add for now, I want quick and easy.
